core/nightshade/src/nightshade.rs

```diff
@@ -0,0 +1,781 @@
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.bare_state.cmp(&other.bare_state)
+    }
+}
+
+impl Eq for State {}
+
+fn merge(state0: &State, state1: &State) -> State {
+    let mut max_state = max(state0, state1).clone();
+    let min_state = min(state0, state1);
+
+    if max_state.endorses() != min_state.endorses() {
+        if min_state.bare_state.primary_confidence > max_state.bare_state.secondary_confidence {
+            max_state.bare_state.secondary_confidence = min_state.bare_state.primary_confidence;
+            max_state.secondary_proof = min_state.primary_proof.clone();
+        }
+    } else {
+        if min_state.bare_state.secondary_confidence > max_state.bare_state.secondary_confidence {
+            max_state.bare_state.secondary_confidence = min_state.bare_state.secondary_confidence;
+            max_state.secondary_proof = min_state.secondary_proof.clone();
+        }
+    }
+
+    max_state
+}
+
+/// Check when two states received from the same authority are incompatible.
+/// Two incompatible states are evidence of malicious behavior.
+fn incompatible_states(state0: &State, state1: &State) -> bool {
+    let merged = merge(state0, state1);
+    let max_state = max(state0, state1);
+
+    &merged != max_state
+}
+
+/// # Nightshade
+///
+/// Each authority must have one Nightshade instance to compute its state, given updates from
+/// other authorities. It contains the logic of the consensus algorithm.
+pub struct Nightshade {
+    owner_id: AuthorityId,
+    num_authorities: usize,
+    states: Vec<State>,
+    is_adversary: Vec<bool>,
+    best_state_counter: usize,
+    seen_bare_states: HashSet<BareState>,
+    pub committed: Option<BlockHeader>,
+    bls_public_keys: Vec<BlsPublicKey>,
+    bls_owner_secret_key: BlsSecretKey,
+}
+
+impl Nightshade {
+    pub fn new(
+        owner_id: AuthorityId,
+        num_authorities: usize,
+        block_header: BlockHeader,
+        bls_public_keys: Vec<BlsPublicKey>,
+        bls_owner_secret_key: BlsSecretKey,
+    ) -> Self {
+        assert_eq!(owner_id, block_header.author);
+        let mut states = vec![];
+
+        for a in 0..num_authorities {
+            if a == owner_id {
+                states.push(State::new(a, block_header.hash, &bls_owner_secret_key));
+            } else {
+                states.push(State::empty());
+            }
+        }
+
+        // Mark first authority triplet as seen from the beginning
+        let mut seen_bare_states = HashSet::new();
+        seen_bare_states.insert(states[owner_id].bare_state.clone());
+
+        Self {
+            owner_id,
+            num_authorities,
+            states,
+            is_adversary: vec![false; num_authorities],
+            best_state_counter: 1,
+            seen_bare_states,
+            committed: None,
+            bls_public_keys,
+            bls_owner_secret_key,
+        }
+    }
+
+    /// Current state of the authority
+    pub fn state(&self) -> State {
```

Comment:
Return reference and only clone it in the caller function when it is needed.

Reply:
done